cmds/free/free.go

```diff
@@ -0,0 +1,343 @@
+	// B is bytes
+	B unit = 0
+	// KB is kibibytes
+	KB = 10
+	// MB is mebibytes
+	MB = 20
+	// GB is gibibytes
+	GB = 30
+	// TB is tebibytes
+	TB = 40
+)
+
+var units = [...]string{"B", "K", "M", "G", "T"}
+
+// FreeConfig is a structure used to configure the behaviour of Free()
+type FreeConfig struct {
+	Unit        unit
+	HumanOutput bool
+	ToJSON      bool
+}
+
+// the following types are used for JSON serialization
+type mainMemInfo struct {
+	Total     uint64 `json:"total"`
+	Used      uint64 `json:"used"`
+	Free      uint64 `json:"free"`
+	Shared    uint64 `json:"shared"`
+	Cached    uint64 `json:"cached"`
+	Buffers   uint64 `json:"buffers"`
+	Available uint64 `json:"available"`
+}
+
+type swapInfo struct {
+	Total uint64 `json:"total"`
+	Used  uint64 `json:"used"`
+	Free  uint64 `json:"free"`
+}
+
+// MemInfo represents the main memory and swap space informatio in a structured
+// manner, suitable for JSON encoding.
+type MemInfo struct {
+	Mem  mainMemInfo `json:"mem"`
+	Swap swapInfo    `json:"swap"`
+}
+
+type meminfomap map[string]uint64
+
+// meminfo returns a mapping that represents the fields contained in
+// /proc/meminfo
+func meminfo() (meminfomap, error) {
+	buf, err := ioutil.ReadFile(meminfoFile)
+	if err != nil {
+		return nil, err
+	}
+	return meminfoFromBytes(buf)
+}
+
+// meminfoFromBytes returns a mapping that represents the fields contained in a
+// byte stream with a content compatible with /proc/meminfo
+func meminfoFromBytes(buf []byte) (meminfomap, error) {
+	ret := make(meminfomap, 0)
+	for _, line := range bytes.Split(buf, []byte{'\n'}) {
+		kv := bytes.SplitN(line, []byte{':'}, 2)
+		if len(kv) != 2 {
+			// invalid line?
+			continue
+		}
+		key := string(kv[0])
+		tokens := bytes.SplitN(bytes.TrimSpace(kv[1]), []byte{' '}, 2)
+		if len(tokens) > 0 {
+			value, err := strconv.ParseUint(string(tokens[0]), 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			ret[key] = value
+		}
+	}
+	return ret, nil
+}
+
+// missingRequiredFields checks if any of the specified fields are present in
+// the input map.
+func missingRequiredFields(m meminfomap, fields []string) bool {
+	for _, f := range fields {
+		if _, ok := m[f]; !ok {
+			log.Printf("Missing field '%v'", f)
+			return true
+		}
+	}
+	return false
+}
+
+// humanReadableValue returns a string representing the input value, treated as
+// a size in bytes, interpreted in a human readable form. E.g. the number 10240
+// woud return the string "10 kB"
+func humanReadableValue(value uint64) string {
+	v := value
+	// bits to shift. 0 means bytes, 10 means kB, and so on. 40 is the highest
+	// and it means tB
+	var shift uint
+	for {
+		if shift >= uint(len(units)*10) {
+			// 4 means tebibyte, we don't go further
+			break
+		}
+		if v/1024 < 1 {
+			break
+		}
+		v /= 1024
+		shift += 10
+	}
+	var decimal uint64
+	if shift > 0 {
+		// no rounding. Is there a better way to do this?
+		decimal = ((value - (value >> shift << shift)) >> (shift - 10)) / 100
```

Comment:
I believe this is wrong. The expression is in base 1024 up until the final `/ 100`. It must be converted from base 1024 to base 1000 before the final `/ 100`. The following works better:

`decimal = ((value - (value >> shift << shift)) >> (shift - 10)) * 1000 / 1024 / 100`

For rounding to the nearest number, it would be easier with floats. The Linux free uses a float + snprintf (which does the rounding).

Reply:
right on the decimal part, will fix, thanks.

On the rounding, I intentionally left it out, as I prefer to know that that's the memory that is *at least* used (or free, etc). Do you think it'd be better to round?

Reply:
That reasoning works. In the docstring at the top, can you mention the values are truncated?

Reply:
sure